### ISSUE TITLE
Password prompts during installation can be a nuisance.

### ISSUE THREAD

**Issue by timmypidashev** [cb-linux/breath#130]
First off, this is not an important issue. Audio is priority, among other things. But for future thought, it would be nice to only need to write the password once throughout the installation. Reason being it occurs 3 times during the script, and does not allow for a hands-off install. 
This only becomes more painful on Arch-based distros, because of the login timeout that occurs, stopping the script prematurely. On the other hand, storing someone's root password as a variable to pipe to the shell can also be quite trivial. Whichever way this goes, just something to consider :D

**Comment by RodBarnes** [comment]
I've given some thought to this, also.  Haven't got a solution yet but would really like it to collect all the entry at the beginning then just proceed to build the image with no further user interaction.

**Comment by timmypidashev** [comment]
@RodBarnes yep, especially after restarting my installation several times due to me just not being present to put in the dang password :joy:

**Comment by timmypidashev** [comment]
The best way to go about this if it's considered a nuisance, in my opinion, would be to restructure `setup.sh` in such a way that all of the user prompted information happens relatively early on, or in mostly the same place. It's already in a good position, so restructuring shouldn't be improbable.

**Comment by timmypidashev** [comment]
On a further note, and @MilkyDeveloper has already done a great job of this, but later on we can continue to abstract away specific parts of the script, for example `FEATURES` can be modulated instead of populating setup.sh, etc. This would allow for easier maintenance, and simplify the installation to any future collaborators willing to contribute. This can be overdone though, and keeping the balance for what to simplify and what to leave in its current state can become a challenge.

**Comment by MilkyDeveloper** [comment]
Great suggestion, I forgot about this issue while developing this. Using tools like `passwd` over STDIN is a bit complicated, but still possible.

I'm a bit hesitant over the abstraction of `FEATURES` because the overall `setup.sh` script should give the reader an overview of what's happening, especially with a feature that doesn't even write to the USB.

I think the ultimate version of this project would be coded in something other shell scripting. If you look in the [PostmarketOS installer](https://gitlab.com/postmarketOS/pmbootstrap/-/tree/master/pmb/) (a similar project but for phones), you can see the installer was rewritten in Python. I love Python, but I fear of getting only slighty more readability, especially when considering that we may need to use `os.system()` sometimes.

Low-level languages wouldn't be suited to this project due to my skill level. Ansible might be a good idea. I think I'll ultimately settle for Python, but the users will have to download a lot more libraries since I never want to run a shell command (`os.sytem()`) from it.

**Comment by timmypidashev** [comment]
@MilkyDeveloper Rewriting in python is a good idea, although I have to agree, the biggest problem becomes the readability. Python requires way more to get the same work done in comparison to the somewhat simple bash scripts breath is composed of. I would suggest sticking to shell scripting as a sort of prototype for the projects view altogether, and then make the switch once we flesh out more of the projects current scope. As far as the current issue at hand, I'll take a look at what options are available and make a pull request that tries to keep it simple, avoiding any restructure of `setup.sh`

**Comment by RodBarnes** [comment]
My thoughts:  In a normal "production release" situation, running setup (whatever form it takes) is a one time deal to create the image.  That image is then used once to install to the internal storage.  After that, neither the setup nor the image would be used again by a given user.  With that model, while I'm fine going into some other language (e.g., python), it would have to really provide value-add to be justify the additional effort over the simplicity of a bash script.

**Comment by MilkyDeveloper** [comment]
So, are you saying that the ISO file created should not be reusable? I see this as a feature rather than a encumbrance. Or, are you saying that only I need to run `setup.sh` to build an ISO and redistribute? This would cause licensing issues.

**Comment by RodBarnes** [comment]
No, not at all.  I guess I was thinking the typical user would be someone with one (or maybe a couple) Chromebooks (CBs).  In that instance, the image would have to be created once for each Chromebook (the image is unique by user/password and hostname) and each image applied to its target CB.  I know that @timmypidashev indicated he had a bunch of CBs he wanted to move to Linux.  But with working audio restricted to specific models (right?) and no microphone (yet), I didn't envision this being used to image a large number of CBs.  But I may not be thinking big enough.  Personally, I'd love this to become something that a lot of people could use and move away from Google. :-)  But right now, you are the one with the knowledge and you've indicated your involvement will be limited going forward.

With the above in mind, it occurs to me that a way forward would be to alter the image build to make it generic to the distro -- free of user/password and hostname -- and only have those things that bring uniqueness provided at the time the image is first booted.  (Right now, I use a script that handles localization and other stuff that could include anything else unique to that install -- such as user/password and hostname.  I used to manage a large number of VMs where we imaged them regularly and this is how I handled it there.
